solver_greedy_max.py

```diff
@@ -19,11 +22,33 @@
 
     # Observation table
     otab_gngs = Table.read(tabdir + 'obstab_gngs_20201123.fits')
-
+    night_date = '2018-06-08'
     # Remove inst and disperser columns so that greedy-max doesn't add time for partner cals
     # otab_gngs.remove_columns(['inst', 'disperser'])
 
-    # Load the observations.
+    sites = list(sites_from_column_names(ttab_gngs.colnames))
+
+    # Resource API mock 
+    resource = Resource('/resource_mock/data')
+    resource.connect(['n','s'])
+    
+    fpu = {site: resource.night_info('fpu', site.name.lower(), night_date) for site in sites}
+    fpur = {site: resource.night_info('fpur', site.name.lower(), night_date) for site in sites}
+    grat = {site: resource.night_info('grat', site.name.lower(), night_date) for site in sites}
+    instruments = {site: resource.night_info('instr', site.name.lower(), night_date) for site in sites}
+    lgs = {site: resource.night_info('LGS', site.name.lower(), night_date) for site in sites}
+    modes = {site: resource.night_info('mode', site.name.lower(), night_date) for site in sites}
+    ifus = {'FPU':None,'FPUr':None}
+    ifus['FPU'] = {site: resource.night_info('fpu-ifu', site.name.lower(), night_date) for site in sites}
+    ifus['FPUr'] = {site: resource.night_info('fpur-ifu', site.name.lower(), night_date) for site in sites}
```

Comment:
Again, better to calculate `site.name.lower()` only once at the top:
`site_lower = site.name.lower()`
and then replace these.

The pattern:
```
variable = {site: resource.night_info(name, site_lower, night_date) for site in sites}
```
can definitely be simplified down to a couple lines. I would change this to (dunno if the dict type is str):
```
def _get_info(info_name: str, site: List[Sites], night_date: str) -> Dict[Site, str]:
     return {site: resource.night_info(info_name, site.name.lower(), night_date) for site in sites }
```

and then replace all of the calls with, for example:
```
fpu = _get_info('fpu', sites, night_date)
```
to avoid the repetition and make it more readable.

Reply:
It's a great idea. The return type might be a bit complicated because it depends on what information is retrieved. Do you know how to handle this?

Reply:
Sure. If you have possible multiple return types, you can use `Either`, like `Either[str, int, Observation]`.